pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java

```diff
@@ -174,6 +166,16 @@ public synchronized Throwable fillInStackTrace() {
         }
     }
 
+    public static Throwable unwrapCompletionException(Throwable ex) {
+        if (ex instanceof CompletionException) {
+            return ex.getCause();
+        } else if (ex instanceof ExecutionException) {
+            return ex.getCause();
```

Comment:
Why cancel recursion?

Reply:
Because of source code by ``CompletableFuture``:

https://github.com/openjdk/jdk/blob/739769c8fc4b496f08a92225a12d07414537b6c0/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#L323-L348

He won't be wrapped twice.

Reply:
Because of source code for ``CompletableFuture``

https://github.com/openjdk/jdk/blob/739769c8fc4b496f08a92225a12d07414537b6c0/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#L323-L348

The ``CompletionException`` won't be wrapped twice.